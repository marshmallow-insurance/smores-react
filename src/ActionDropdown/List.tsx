import React, { FC } from 'react'
import styled from 'styled-components'

import { Color, theme } from '../theme'

export type ActionListItem = {
  label: string
  value: string
  bgColor?: Color
  textColor?: Color
}

type Props = {
  options: ActionListItem[]
  selectOption: (option: ActionListItem) => void
}

export const List: FC<Props> = ({ options, selectOption }) => (
  <Container>
    {options.map((option) => (
      <ListItem key={option.value} onClick={() => selectOption(option)}>
        {option.label}
      </ListItem>
    ))}
  </Container>
)

const Container = styled.ul`
  padding: 0;
  margin: 0;
  max-height: 188px;
`

const ListItem = styled.li`
  padding: 16px 24px;
  text-align: left;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  box-sizing: border-box;
  font-size: 14px;
  cursor: pointer;
  color: ${theme.colors.liquorice};
  border-bottom: 1px solid ${theme.colors.chia};

  &:hover {
    background-color: ${theme.colors.coconut};
  }

  &:last-child {
    border-bottom: none;
  }
`
