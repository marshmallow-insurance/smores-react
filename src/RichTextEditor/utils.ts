import { $createParagraphNode, LexicalNode, RootNode } from 'lexical'

export const appendNodes = (root: RootNode, nodes: LexicalNode[]) => {
  nodes
    .filter((node) => node.__type !== 'linebreak')
    .map((node) => {
      if (node.__type === 'text') {
        const paragraphNode = $createParagraphNode()
        paragraphNode.append(node)
        return paragraphNode
      }
      return node
    })
    .forEach((node) => root.append(node))
}
