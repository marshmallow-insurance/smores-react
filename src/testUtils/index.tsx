import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
