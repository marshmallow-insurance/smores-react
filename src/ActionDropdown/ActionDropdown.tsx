import React, { FC, useState } from 'react'
import styled, { css } from 'styled-components'

import { Text } from '../Text'
import { Icon } from '../Icon'
import { List, ActionListItem } from './List'

import { theme } from '../theme'

type Props = {
  /** ID, usually used for tests  */
  id: string
  /** className attribute to apply classes from props */
  className?: string
  /** label displayed above the dropdown  */
  label?: string
  /** value */
  value: ActionListItem
  /** list of items for the dropdown list */
  list: ActionListItem[]
  /** onSelect handler */
  onSelect: (action: ActionListItem) => void
}

export const ActionDropdown: FC<Props> = ({
  id,
  className = '',
  label,
  value,
  list,
  onSelect,
}) => {
  const [open, setOpen] = useState(false)

  return (
    <Container
      id={id}
      className={className}
      onClick={() => setOpen((currentOpen) => !currentOpen)}
    >
      {label && (
        <Text tag="label" color="grey4" typo="label">
          {label}
        </Text>
      )}

      <Label text={value.textColor ?? 'blue7'} bg={value.bgColor ?? 'grey2'}>
        <SelectedOption>{value.label}</SelectedOption>
        <Icon
          render="caret"
          color={value.textColor ?? 'grey7'}
          size={24}
          rotate={open ? 180 : 0}
        />
      </Label>
      <OuterContainer open={open}>
        <List options={list} selectOption={onSelect} />
      </OuterContainer>

    </Container>
  )
}

interface ISelected {
  text: string
  bg: string
}

const Label = styled.div<ISelected>(
  ({ text, bg }) => css`
    color: ${theme.colors[text]};
    background-color: ${theme.colors[bg]};
    position: relative;
    width: 100%;
    height: 48px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    text-align: center;
    border-radius: 8px;
    padding: 16px 16px 14px;
    box-sizing: border-box;
    user-select: none;
  `,
)

const SelectedOption = styled.span`
  font-size: 14px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`

const Container = styled.div`
  position: relative;
  display: block;
  width: 210px;
  height: 48px;
  text-align: center;
  cursor: pointer;
  outline: none;
  border-radius: 8px;
`

interface IOpen {
  open: boolean
}

const OuterContainer = styled.div<IOpen>(
  ({ open }) => css`
    position: absolute;
    width: 100%;
    border-radius: 8px;
    margin-top: 8px;
    display: ${open ? 'block' : 'none'};
    max-height: ${open ? '235px' : '48px'};
    background-color: ${theme.colors.white};
    border: 1px solid ${theme.colors.grey3};
    overflow-y: ${open ? 'auto' : 'hidden'};
    z-index: 2;
    transition: all 0.2s ease-in-out;
  `,
)