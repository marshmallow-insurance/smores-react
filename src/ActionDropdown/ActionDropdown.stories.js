import React from 'react'
import { ActionDropdown } from './ActionDropdown'
import { Container as ActionDropdownContainer } from './Container'

export default {
  title: 'Action Dropdown',
  component: ActionDropdown,
}

const Template = (args) => <ActionDropdownContainer {...args} />

export const Default = Template.bind({})

Default.args = {
  textColor: 'secondary',
}
