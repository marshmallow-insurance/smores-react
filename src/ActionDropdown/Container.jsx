import React, { useState } from 'react'
import { Box } from '../Box'
import { ActionDropdown } from './ActionDropdown'
import { Text } from '../Text'

const actions = [
  {
    label: 'Default grey',
    value: 'DEFAULT',
    bgColor: 'grey2',
    textColor: 'blue7',
  },
  {
    label: 'Medium',
    value: 'MEDIUM',
    bgColor: 'yellow5',
    textColor: 'blue7',
  },
  {
    label: 'High alert',
    value: 'HIGH_ALERT',
    bgColor: 'orange4',
    textColor: 'white',
  },
  {
    label: 'Good zone',
    value: 'GOOD_ZONE',
    bgColor: 'green6',
    textColor: 'white',
  },
  {
    label: 'DANGER zone',
    value: 'DANGER_ZONE',
    bgColor: 'red6',
    textColor: 'white',
  },
]

export const Container = () => {
  const [selectedAction, setSelectedAction] = useState(actions[0])

  return (
    <Box flex alignItems="center">
      <ActionDropdown
        id="statusActions"
        list={actions}
        value={selectedAction}
        onSelect={setSelectedAction}
      />

      <Box mr="16px" />

      <Box>
        <Text tag="p" typo="base-small">
          {selectedAction.label}
        </Text>
      </Box>
    </Box>
  )
}
