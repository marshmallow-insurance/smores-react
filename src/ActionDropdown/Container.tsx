import React, { useState } from 'react'
import { Box } from '../Box'
import { ActionDropdown, ActionDropdownProps } from './ActionDropdown'
import { Text } from '../Text'

const actions: ActionDropdownProps['list'] = [
  {
    label: 'Default grey',
    value: 'DEFAULT',
    bgColor: 'coconut',
    textColor: 'liquorice',
  },
  {
    label: 'Brand',
    value: 'ACTION',
    bgColor: 'marshmallowPink',
    textColor: 'liquorice',
  },
  {
    label: 'Medium',
    value: 'MEDIUM',
    bgColor: 'lemon',
    textColor: 'liquorice',
  },
  {
    label: 'High alert',
    value: 'HIGH_ALERT',
    bgColor: 'tangerine',
    textColor: 'cream',
  },
  {
    label: 'Good zone',
    value: 'GOOD_ZONE',
    bgColor: 'apple',
    textColor: 'cream',
  },
  {
    label: 'DANGER zone',
    value: 'DANGER_ZONE',
    bgColor: 'strawberry',
    textColor: 'cream',
  },
]

export const Container = () => {
  const [selectedAction, setSelectedAction] = useState(actions[0])

  return (
    <Box flex alignItems="center">
      <ActionDropdown
        id="statusActions"
        list={actions}
        value={selectedAction}
        onSelect={setSelectedAction}
      />

      <Box mr="16px" />

      <Box>
        <Text tag="p" typo="base-small">
          {selectedAction.label}
        </Text>
      </Box>
    </Box>
  )
}
