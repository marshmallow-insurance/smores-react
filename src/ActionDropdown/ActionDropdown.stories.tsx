import React from 'react'
import { ActionDropdown } from './ActionDropdown'
import { Container as ActionDropdownContainer } from './Container'

export default {
  title: 'Action Dropdown',
  component: ActionDropdown,
}

const Template = () => <ActionDropdownContainer />

export const Default = Template.bind({})

Default.args = {
  textColor: 'liquorice',
}
