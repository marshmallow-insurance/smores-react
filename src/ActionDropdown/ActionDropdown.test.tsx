import React from 'react'
import { render } from '../testUtils'
import { ActionDropdown } from './ActionDropdown'
import { ActionListItem } from './List'
import { noop } from '../utils/noop'

const options: ActionListItem[] = [
  {
    label: 'Default grey',
    value: 'DEFAULT',
    bgColor: 'color.surface.base.100',
    textColor: 'color.text.base',
  },
  {
    label: 'Brand',
    value: 'ACTION',
    bgColor: 'color.surface.brand.300',
    textColor: 'color.text.base',
  },
]

describe('ActionDropdown', () => {
  it('renders correctly with default props', () => {
    const { container } = render(
      <ActionDropdown
        value={{ label: 'Default grey', value: 'DEFAULT' }}
        list={options}
        onSelect={noop}
      />,
    )
    expect(container).toMatchSnapshot()
  })
})
