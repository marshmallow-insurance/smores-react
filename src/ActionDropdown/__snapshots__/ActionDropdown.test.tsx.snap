// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionDropdown > renders correctly with default props 1`] = `
<div>
  <div
    class="sc-bRKDuR bHjpxP sc-kNOymR iMmesL"
  >
    <div
      class="sc-jJLAfE kVjsTn"
    >
      <span
        class="sc-hwkwBN bjyVWn"
      >
        Default grey
      </span>
      <span
        class="sc-bRKDuR bHjpxP sc-fhHczv XmxEI"
        data-testid="caret-container"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="sc-dYwGCk dvdcAq"
    >
      <ul
        class="sc-jwTyAe gMUwNg"
      >
        <li
          class="sc-hjsuWn dRRnwj"
        >
          Default grey
        </li>
        <li
          class="sc-hjsuWn dRRnwj"
        >
          Brand
        </li>
      </ul>
    </div>
  </div>
</div>
`;
